setup/tue-functions.bash

```diff
@@ -19,6 +19,166 @@ function _list_subdirs
+{
+    local default_branch
+    default_branch=$(git remote show origin | grep HEAD | awk '{print $3}')
+    _git_remote_checkout origin "$default_branch"
+}
+
+function _tue-default-branch
+{
+    _tue-repos-do "__tue-default-branch"
+}
+
+function _tue-git-branch-clean
+{
+    # Function to remove stale branches from a git repository (which should
+    # either be the PWD or one of its parent directories). The function removes
+    # stale branches in two layers. First it removes all branches that have been
+    # merged in the remote, then it checks for unmerged branches that have been
+    # deleted from the remote and prompts for confirmation before removal. If
+    # the function is called with "--force-remove" flag, then no confirmation is asked
+
+    local force_remove=
+    local error_code=
+    local stale_branches=
+    local repo=
+    repo="$PWD"
+
+    if [ -n "$1" ]
+    then
+        if [ "$1" == "--force-remove" ]
+        then
+            force_remove=true
+        else
+            echo -e "[tue-git-branch-clean] Error! Unknown input argument '$1'. Only supported argument is '--force-remove' to forcefully remove unmerged stale branches"
+            return 1
+        fi
+    fi
+
+    git fetch -p || { echo -e "[tue-git-branch-clean] 'git fetch -p' failed in '$repo'."; return 1; }
+
+    stale_branches=$(git branch --list --format "%(if:equals=[gone])%(upstream:track)%(then)%(refname)%(end)" \
+| sed 's,^refs/heads/,,;/^$/d')
+
+    if [ -z "$stale_branches" ]
+    then
+        echo -e "[tue-git-branch-clean] No branches to remove in repository '$repo'. Cleanup complete."
+        return 0
+    fi
+
+    # If the current branch is a stale branch then change to the default repo
+    # branch before cleanup
+    if [[ "$stale_branches" == *$(git rev-parse --abbrev-ref HEAD)* ]]
+    then
+        __tue-robocup-default-branch
+
+        git pull --ff-only --prune > /dev/null 2>&1
+        error_code=$?
+
+        if [ ! $error_code -eq 0 ]
+        then
+            echo -e "[tue-git-branch-clean] Error pulling upstream on default branch of repository '$repo'. Cancelling branch cleanup."
+            return 1
+        fi
+    fi
+
+    echo -e "Removing branches:"
+    echo -e "------------------"
+    echo -e "$stale_branches"
+
+    local stale_branch=
+    local unmerged_stale_branches=
+    for stale_branch in $stale_branches
+    do
+        git branch -d "$stale_branch" > /dev/null 2>&1
+        error_code=$?
+
+        # If an error occured in safe deletion of a stale branch, add it to the
+        # list of unmerged stale branches which are to be forcefully removed
+        # upon confirmation by the user
+        if [ ! $error_code -eq 0 ]
+        then
+            unmerged_stale_branches="$unmerged_stale_branches $stale_branch"
+        fi
+    done
+
+    # Removal of unmerged stale branches. Not a default operation with the high
+    # level command tue-git-branch-clean
+    if [ -n "$unmerged_stale_branches" ]
```

Comment:
I would be annoyed by this. But I don't want to force remove branches.

Reply:
That's the only way of removing unmerged stale branches. Safe deletion gives an error with such branches. And just for safety I ask for user confirmation. In practice you will see that only few repos will have such branches so the user won't be prompted so many times.

And if the user knows what he's doing, then he can call the function with `--force-remove` to remove the unmerged stale branches

Reply:
I think I have a nice solution that wouldn't be annoying. If there are unmerged stale branches, I'll just print out the names of the branches and instead of asking for user input, I'll echo to remove these stale branches, call the command with `--force-remove`.

What do you think about this?

Reply:
Did you implement this?

Reply:
Yup. I think you know it by now :laughing: